File: automate.py
#Script that automates the setup for kerberossdr, change variable values at top of code to what inputs you will use.
#Plug in your terminals before starting script and switch them out with your antennas after it runs.
#Author: Yehia Helwa
#Created: 7/1/2020

from selenium import webdriver
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import Select
from selenium.webdriver.chrome.options import Options
import time

#input variables here
#make sure that all your inputs are valid entries on the webpage otherwise you will get an error thrown back, values listed below confirmed to work
centerFreq='433.3'
sampFreq='1.024'
gain='28.0'
filtBW='10'
firTap='100'
decimation='1'
antArrang='UCA'
spacing='0.24' #spacing in meters




#selecting chromium browser
waitDelay=10
options= Options()
browser=webdriver.Chrome(options=options, executable_path='/usr/bin/chromedriver')
type(browser)

#go to config page
browser.get('http://192.168.0.120:8080/init')


centerFreqElem= WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "center_freq")))
centerFreqElem.clear()
centerFreqElem.send_keys(centerFreq)
print("Inputted center frequency with value <%s>" %centerFreq)

sampFreqElem= WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "samp_freq")))
select = Select(sampFreqElem)
select.select_by_visible_text(sampFreq)
print("Inputted sample frequency with value <%s>" %sampFreq)

uniGainElem = WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.ID, "uniform_gain_id")))
if not uniGainElem.is_selected():
 uniGainElem.click()
print("Enabled uniform gain")

gainElem= WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "gain")))
select = Select(gainElem)
select.select_by_visible_text(gain)
print("Selected gain with value <%s>" %gain)


#update receiver params
updateReciever = WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.XPATH, '/html/body/div[2]/form/div[9]/input')))
updateReciever.click()
print("Updated receiver parameters")
time.sleep(2)

dcCompElem =  WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "dc_comp")))
if not dcCompElem.is_selected():
 dcCompElem.click()
print("Enabled DC compensation")

filtBwElem=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.XPATH, '/html/body/div[3]/form/div[3]/div[2]/input')))
filtBwElem.clear()
filtBwElem.send_keys(filtBW)

firTapElem= WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "fir_size")))
firTapElem.clear()
firTapElem.send_keys('0')

decimationElem= WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "decimation")))
decimationElem.clear()
decimationElem.send_keys('1')
print("Set syncing IQ parameters")


#update IQ params
updateIQ = WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.XPATH, '/html/body/div[3]/form/div[6]/input')))
updateIQ.click()
print("Updated IQ parameters")
time.sleep(2)

#start processing
startProc=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.XPATH, '/html/body/div[4]/form[1]/div/input')))
startProc.click()
print("Started processsing")
time.sleep(2)

#start spectrum display
# browser.find_element_by_xpath('/html/body/div[5]/form[1]/div/input').click()	
# time.sleep(0.5)

#switch to sync
browser.get('http://192.168.0.120:8080/sync')
print("Switching to sync tab")


#enable noise source and sync display
syncButton=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.XPATH, '/html/body/div[2]/form[1]/div/input')))
syncButtonTxt=syncButton.get_attribute("value")
if(syncButtonTxt=="Enable Noise Source & Sync Display"):
 print("NS and SD disabled now enabling")
 syncButton.click()
elif(syncButtonTxt=="Disable Noise Source & Sync Display"):
 print("NS and SD enabled (not supposed to happen)")
time.sleep(4)

#sample sync then calibrate calibrate iq
sampleSync=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.XPATH, '/html/body/div[3]/form[2]/div/input')))
sampleSync.click()	
print("Syncing system ...")
time.sleep(6)


calibIQ=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.XPATH, '/html/body/div[3]/form[3]/div/input')))
calibIQ.click()	
print("Calibrating IQ ...")
time.sleep(6)

#disable noise source and sync display
syncButtonTxt=browser.find_element_by_xpath("/html/body/div[2]/form[1]/div/input").get_attribute("value")
if(syncButtonTxt=="Enable Noise Source & Sync Display"):
 print("NS and SD disabled (not supposed to happen)")
elif(syncButtonTxt=="Disable Noise Source & Sync Display"):
 print("NS and SD enabled now disabling (syncing done)")
 browser.find_element_by_xpath("/html/body/div[2]/form[1]/div/input").click()
time.sleep(3)

#switch back config to update FIR tap and decimation
browser.get('http://192.168.0.120:8080/init')
print("Switching to config tab to update IQ parameters")


dcCompElem = WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "dc_comp")))
if not dcCompElem.is_selected():
 dcCompElem.click()

filtBwElem=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.XPATH, '/html/body/div[3]/form/div[3]/div[2]/input')))
filtBwElem.clear()
filtBwElem.send_keys(filtBW)

firTapElem= WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "fir_size")))
firTapElem.clear()
firTapElem.send_keys(firTap)

decimationElem= WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "decimation")))
decimationElem.clear()
decimationElem.send_keys(decimation)
print("Inputted actual IQ parameters")


#update IQ params

updateIQ = WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.XPATH, '/html/body/div[3]/form/div[6]/input')))
updateIQ.click()
print("Updated IQ parameters")
time.sleep(2)


#switch to doa
browser.get('http://192.168.0.120:8080/doa')
print("Switching to DOA tab")


#select UCA arrangement
antArrangElement= WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "ant_arrangement")))
select = Select(antArrangElement)
select.select_by_visible_text(antArrang)
print("Selected <%s> arrangement" %antArrang)


#set spacing
antSpacing= WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "ant_spacing")))
antSpacing.clear()
antSpacing.send_keys(spacing)
print("Setting spacing of <%s> meters" %spacing)


#enable DOA
enDOA = WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "en_doa")))
if not enDOA.is_selected():
 enDOA.click()

#disable all algorithims except MUSIC
enBartlett=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "en_bartlett")))
enCapon=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "en_capon")))
enMEM=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "en_MEM")))
enMUSIC=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.NAME, "en_MUSIC")))
if enBartlett.is_selected():
	enBartlett.click()

if enCapon.is_selected():
	enCapon.click()

if enMEM.is_selected():
	enMEM.click()

if not enMUSIC.is_selected():
	enMUSIC.click()

print("Enabling DOA and disabling all algorithims except MUSIC")
time.sleep(2)

updateDOA=WebDriverWait(browser, waitDelay).until(EC.presence_of_element_located((By.XPATH, '/html/body/div[3]/form/div[6]/input')))
updateDOA.click()
print("DOA enabled!")
time.sleep(5)
print("Shutting down browser")
browser.quit()

File: automate.js
let puppeteer = require('puppeteer');



(async () => {
  var os = require('os');
  var ifaces = os.networkInterfaces();
  var addy
  Object.keys(ifaces).forEach(function (ifname) {
    var alias = 0;

    ifaces[ifname].forEach(function (iface) {
      if ('IPv4' !== iface.family || iface.internal !== false) {
        // skip over internal (i.e. 127.0.0.1) and non-ipv4 addresses
        return;
      }

      if (alias >= 1) {
        // this single interface has multiple ipv4 addresses
        console.log(ifname + ':' + alias, iface.address);
      } else {
        // this interface has only one ipv4 adress
      
	addy=iface.address
      }
      ++alias;
    });
  });


  console.log('IP address: '+addy)
  var http = 'http://'
  var goto  
  let browser = await puppeteer.launch({executablePath: '/usr/bin/chromium-browser'});
  let page = await browser.newPage();
  await page.goto(http.concat(addy,':8080/init'));
  await page.waitForSelector('.btn');
//CHANGE THESE VARS
  let centerFreq = '549';
  let sampFreq = '2';//The number refers to the index of the dropdown starting at 0, ex. '2' refers to the third choice from dropdown
  let gain = '15';//Use same index principle
  let filtBW = '200';
  let firTap = '100';
  let decimation = '1';
  let ant = '1';//Also index
//This is automatic calculating for the spacing using 0.333 spacing factor
  let calc = (299792458/(Number(centerFreq)*(10**6)))/3;
  let rcalc= calc.toFixed(4);
  let spacing = rcalc.toString();//if you want to use fixed value just add it here ex '0.14'(unit is meters)
  console.log('Spacing is '+spacing+' meters');
  let tDelay = 100 ;

  let centerFreqElem = await page.$('[name="center_freq"]');
  let sampFreqElem = 'select[name="samp_freq"]';
  let uniGainElem = await page.$('[name="uniform_gain"]');
  let gainElem = 'select[name="gain"]';
  let updateRecElem = await page.$('[value="Update Receiver Paramaters"]');

  
  

  await centerFreqElem.click({clickCount:3});
  await centerFreqElem.type(centerFreq);

  await page.select(sampFreqElem, sampFreq);

  if(!(await (await uniGainElem.getProperty('checked')).jsonValue())){
	await uniGainElem.click();  }

  await page.select(gainElem, gain);
   

//Updates receiver params
  await updateRecElem.click();
  console.log("Updated Receiver Paramaters")
  await page.waitFor(tDelay);  
  await page.goto(http.concat(addy,':8080/init'));
  await page.waitForSelector('.btn');

  let dcCompElem = await page.$('[name="dc_comp"]');
  let filtBWElem = await page.$('[name="filt_bw"]');
  let firTapElem = await page.$('[name="fir_size"]');
  let decimationElem = await page.$('[name="decimation"]');
  let updateIQElem = await page.$('[value="Update IQ Paramaters"]');

  
  if(!(await (await dcCompElem.getProperty('checked')).jsonValue())){
	await dcCompElem.click();  }
	
  await filtBWElem.click({clickCount:3});
  await filtBWElem.type(filtBW);
 
  await firTapElem.click({clickCount:3});
  await firTapElem.type('0');
 
  await decimationElem.click({clickCount:3});
  await decimationElem.type('1');
  await updateIQElem.click();
  console.log("Updated IQ with sync paramaters")
  await page.waitFor(tDelay);

  await page.goto(http.concat(addy,':8080/init'));
  await page.waitForSelector('.btn');
  let startProcessingElem = await page.$('[value="Start Processing"]');

  await startProcessingElem.click();
  await page.waitFor(tDelay);
  await page.goto(http.concat(addy,':8080/sync'));
  await page.waitForSelector('.btn');
  let enSyncElem = await page.$('[value="Enable Noise Source & Sync Display"]');
  let disableSyncElem = await page.$('[value="Disable Noise Source & Sync Display"]');

  try { await enSyncElem.click();}
  catch (error) { console.log("Noise source already enabled");}
  await page.waitFor(tDelay);
  await page.goto(http.concat(addy,':8080/sync'));
  await page.waitForSelector('.btn');
  let sampleSyncElem = await page.$('[value="Sample Sync"]');
  await page.waitFor(3000);
  await sampleSyncElem.click();	
  console.log("Beginning Sample Sync");
  await page.waitFor(tDelay);
  await page.goto(http.concat(addy,':8080/sync'));
  await page.waitFor(10000);
  let calibIQ = await page.$('[value="Calibrate IQ"]');
  await calibIQ.click();
  console.log("Beginning IQ Calibration");
  await page.waitFor(10000);
  
  await page.goto(http.concat(addy,':8080/sync'));
  await page.waitForSelector('.btn');
  enSyncElem = await page.$('[value="Enable Noise Source & Sync Display"]');
  disableSyncElem = await page.$('[value="Disable Noise Source & Sync Display"]');

  try { await disableSyncElem.click();}
  catch (error) { console.log("Noise source already off");}
  console.log("Syncing done, disabline noise source")
  await page.waitFor(tDelay); 
  await page.goto(http.concat(addy,':8080/init'));
  await page.waitForSelector('.btn');

  dcCompElem = await page.$('[name="dc_comp"]');
  filtBWElem = await page.$('[name="filt_bw"]');
  firTapElem = await page.$('[name="fir_size"]');
  decimationElem = await page.$('[name="decimation"]');
  updateIQElem = await page.$('[value="Update IQ Paramaters"]');

  
  if(!(await (await dcCompElem.getProperty('checked')).jsonValue())){
	await dcCompElem.click();  }
	
  await filtBWElem.click({clickCount:3});
  await filtBWElem.type(filtBW);
 
  await firTapElem.click({clickCount:3});
  await firTapElem.type(firTap);
 
  await decimationElem.click({clickCount:3});
  await decimationElem.type(decimation);
  await updateIQElem.click();
  console.log("Updated IQ with actual paramaters")
  await page.waitFor(tDelay);

  
  await page.goto(http.concat(addy,':8080/doa'));
  await page.waitForSelector('.btn');
  let antElem = 'select[name="ant_arrangement"]';
  let spacingElem = await page.$('[name="ant_spacing"]');
  let enDOAElem = await page.$('[name="en_doa"]');
  let enBartElem = await page.$('[name="en_bartlett"]');
  let enCaponElem = await page.$('[name="en_capon"]');
  let enMEMElem = await page.$('[name="en_MEM"]');
  let enMUSICElem = await page.$('[name="en_MUSIC"]');
  let updateDOAElem = await page.$('[value="Update DOA"]');

  await page.select(antElem, ant);
  await spacingElem.click({clickCount:3});
  await spacingElem.type(spacing);
  if(!(await (await enDOAElem.getProperty('checked')).jsonValue())){
	await enDOAElem.click();  }
  if((await (await enBartElem.getProperty('checked')).jsonValue())){
	await enBartElem.click();  }
  if((await (await enCaponElem.getProperty('checked')).jsonValue())){
	await enCaponElem.click();  }
  if((await (await enMEMElem.getProperty('checked')).jsonValue())){
	await enMEMElem.click();  }
  if(!(await (await enMUSICElem.getProperty('checked')).jsonValue())){
	await enMUSICElem.click();  }
  await updateDOAElem.click();
  console.log("Updated DOA parameters");
  await page.waitFor(tDelay);
  console.log("Closing browser");
//  await page.screenshot({path: 'screenshot.png'});
  await browser.close();
})();
